### DISCUSSION TITLE
Powershell module? (Feature Request)

### DISCUSSION THREAD

**Discussion by ilgrank** [dotnet/MQTTnet#1697] [Ideas]
Hi
[as discussed here](https://github.com/dotnet/MQTTnet/issues/1330#issuecomment-1407444758) , the current release is, _de facto_ , not fully usable in powershell due to powershell being unable to handle async events.

Since @chkr1011 mentioned the possibility of making a dedicated powershell module, Is there any chance (please! 🙏 ) that this will go on roadmap?

Thanks :)

**Comment by chkr1011** [comment]
I did not started yet due to lack of time but this is still a topic for the future.

**Reply by ilgrank** [reply]
High hopes for that, keeping fingers crossed then, thanks! :)

**Reply by vireshov** [reply]
+1 ! @chkr1011 :)

**Comment by vireshov** [comment]
Hi @chkr1011,

Thanks for considering this feature and thanks @ilgrank  for making the request.

Definitely a +1 for this.

We are looking to use the MQTTnet library as a client to interact both ways with VMware VCD's MQTT broker for our own custom automation usecases. In the VMware "universe", PowerShell is the most common used automation language for automation purposes. Infra Engineers can automate, but they are not full fledged developers. Hence PowerShell has a strong use case.

As we need this functionality today, we are now going to attempt to create a simple lightweight wrapper around the MQTTnet Library, for the purposes we need it for...but a PS Module developed by you guys would be awesome.
We use PS7 on Linux, but I think this library has dotnet core support, right?

Is there any WIP and/or ETA for this feature?

Thank you for your attention to this matter !

**Comment by chkr1011** [comment]
Since we got AI these days I gave it a try even if I am absolutely no PowerShell expert but I created a branch and PR (https://github.com/dotnet/MQTTnet/pull/2215) which contains a very basic PowerShell module. Please let me know what you think. My next steps are exposing more Parameters and exposing several event handlers. There is also a Test.ps1 file which shows how to use the module.

**Reply by ilgrank** [reply]
Hi!
Thanks for your effort!
I have tried checking out the powershell branch, but could not find any compiled DLL to try it. Could you please build one? 
Many thanks

**Reply by chkr1011** [reply]
If you checked it out you only need to recompile the solution. Then you can execute the Test.ps1 powershell file. If you do not have any IDE and cannot compile the code I will try to embed it in the repo (for testing purposes only).

**Reply by ilgrank** [reply]
Thanks
I'm more of a scripting guy, and if you can provide a test DLL that would help greatly

**Reply by chkr1011** [reply]
I pushed the DLL. Let me know if it works.

**Reply by ilgrank** [reply]
> I pushed the DLL. Let me know if it works.

Hi! Thanks for compiling the DLL! 
I've tried it an so far, both connecting and posting do work perfectly!

The only problem I'm having is... you guessed it probably, with the message received event 
It's not erroring, but neither I'm receiving any event (probably my fault):

```
Import-Module ./bin/Debug/net8.0/MQTTnet.PowerShell.dll -verbose
$session = New-MqttSession
Connect-MqttSession -Session $session -Host "192.168.0.102" -username "user01" -password "password01"
Publish-MqttMessage -Session $session  -Topic "Test" -Payload "Hello from PowerShell"
Subscribe-MqttTopic -Session $session -Topic "OpenMQTTGateway/BTtoMQTT/*"

#Receive-MqttMessage -Session $session -TimeoutSeconds 20
#Disconnect-MqttSession -Session $session
#Remove-MqttSession -Session $session

Register-ObjectEvent -InputObject $session -EventName MessageReceived -Action {
        $topic = $EventArgs.Topic
        $payload = $EventArgs.Payload
        $qos = $EventArgs.QoS
        $retain = $EventArgs.Retain
        $userProps = $EventArgs.UserProperties
        
        Write-Host "Received message: Topic=$topic, Payload=$payload, QoS=$qos, Retain=$retain"
        if ($userProps.Count -gt 0) {
            Write-Host "User Properties:"
            foreach ($prop in $userProps.GetEnumerator()) {
                Write-Host "  $($prop.Key): $($prop.Value)"
            }
        }
    }
```
and I'm stuck here since a couple of days, and running out of ideas..

**Reply by ilgrank** [reply]
Hello @chkr1011 , still here and willing to cooperate if I can be of any help in testing, thanks :)
